boot/seal.go

```diff
@@ -518,56 +518,69 @@ func resealFallbackObjectKeys(pbc predictableBootChains, authKeyFile string, rol
-		// get kernel and gadget information from seed
-		perf := timings.New(nil)
-		_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, system, []snap.Type{snap.TypeKernel, snap.TypeGadget}, perf)
-		if err != nil {
-			return nil, fmt.Errorf("cannot read system %q seed: %v", system, err)
-		}
-		if len(snaps) != 2 {
-			return nil, fmt.Errorf("cannot obtain recovery system snaps")
-		}
-		seedKernel, seedGadget := snaps[0], snaps[1]
-		if snaps[0].EssentialType == snap.TypeGadget {
-			seedKernel, seedGadget = seedGadget, seedKernel
-		}
+	chainsForModel := func(model secboot.ModelForSealing) error {
+		for _, system := range systems {
+			// get kernel and gadget information from seed
+			perf := timings.New(nil)
+			_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, system, []snap.Type{snap.TypeKernel, snap.TypeGadget}, perf)
+			if err != nil {
+				return fmt.Errorf("cannot read system %q seed: %v", system, err)
+			}
+			if len(snaps) != 2 {
+				return fmt.Errorf("cannot obtain recovery system snaps")
+			}
+			seedKernel, seedGadget := snaps[0], snaps[1]
+			if snaps[0].EssentialType == snap.TypeGadget {
+				seedKernel, seedGadget = seedGadget, seedKernel
+			}
 
-		// get the command line
-		cmdline, err := composeCommandLine(currentEdition, ModeRecover, system, seedGadget.Path)
-		if err != nil {
-			return nil, fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
-		}
+			// get the command line
+			cmdline, err := composeCommandLine(currentEdition, ModeRecover, system, seedGadget.Path)
+			if err != nil {
+				return fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+			}
 
-		var kernelRev string
-		if seedKernel.SideInfo.Revision.Store() {
-			kernelRev = seedKernel.SideInfo.Revision.String()
-		}
+			var kernelRev string
+			if seedKernel.SideInfo.Revision.Store() {
+				kernelRev = seedKernel.SideInfo.Revision.String()
+			}
 
-		recoveryBootChain, err := trbl.RecoveryBootChain(seedKernel.Path)
-		if err != nil {
-			return nil, err
+			recoveryBootChain, err := trbl.RecoveryBootChain(seedKernel.Path)
+			if err != nil {
+				return err
+			}
+
+			// get asset chains
+			assetChain, kbf, err := buildBootAssets(recoveryBootChain, modeenv)
+			if err != nil {
+				return err
+			}
+
+			chains = append(chains, bootChain{
+				BrandID:        model.BrandID(),
+				Model:          model.Model(),
+				Grade:          model.Grade(),
+				ModelSignKeyID: model.SignKeyID(),
+				AssetChain:     assetChain,
+				Kernel:         seedKernel.SnapName(),
+				KernelRevision: kernelRev,
+				KernelCmdlines: []string{cmdline},
+				kernelBootFile: kbf,
+			})
 		}
+		return nil
+	}
 
-		// get asset chains
-		assetChain, kbf, err := buildBootAssets(recoveryBootChain, modeenv)
-		if err != nil {
+	if err := chainsForModel(modeenv.ModelForSealing()); err != nil {
+		return nil, err
+	}
+
+	if modeenv.TryModel != "" {
+		if err := chainsForModel(modeenv.TryModelForSealing()); err != nil {
```

Comment:
should we incorporate the try model in the fallback/recovery keys or only the run keys, like we do for the try systems?

Reply:
@bboozzoo maybe we should discuss a bit because the try model exists/is set very late, and the relation there with the main model is almost reversed?